package com.lvl6.proto;

option java_outer_classname = "InfoProto";

//------------------------------------------------------------------------------------------

enum UserType {
	GOOD_WARRIOR = 0;
	GOOD_ARCHER = 1;
	GOOD_MAGE = 2;
	BAD_WARRIOR = 3;
	BAD_ARCHER = 4;
	BAD_MAGE = 5;
}

message MinimumUserProto {
	required int32 userId = 1;
	optional string name = 2;
	optional UserType userType = 3;
}

message FullUserProto {
	required int32 userId = 1;
	required string name = 2;
	required int32 level = 3;
	required UserType userType = 4;
	required int32 attack = 5;
	required int32 defense = 6;
	required int32 stamina = 7;
	required int32 energy = 8;
	required int32 skillPoints = 10;
	required int32 healthMax = 11;
	required int32 energyMax = 12;
	required int32 staminaMax = 13;
	required int32 diamonds = 14;
	required int32 coins = 15;
	required int32 wood = 16;
	required int32 marketplaceDiamondsEarnings = 17;
	required int32 marketplaceCoinsEarnings = 18;
	required int32 marketplaceWoodEarnings = 19;
	required int32 vaultBalance = 20;
	required int32 experience = 21;
	required int32 tasksCompleted = 22;
	required int32 battlesWon = 23;
	required int32 battlesLost = 24;
	required int32 hourlyCoins = 25;
	required string armyCode = 26;
	required int32 numReferrals = 27;
	required string udid = 28;
	required LocationProto userLocation = 29;
	required int32 numPostsInMarketplace = 30;
}

//------------------------------------------------------------------------------------------

message FullEquipProto {
	required int32 equipId = 1;
	required string name = 2;
	required EquipType equipType = 3;
	required ClassType classType = 4;
	required int32 attackBoost = 5;	
	required int32 defenseBoost = 6;
	required int32 minLevel = 7;
	optional int32 coinPrice = 8;
	optional int32 diamondPrice = 9;
	required float chanceOfLoss = 10;
	required Rarity rarity = 11;
	
	enum Rarity {
		COMMON = 0;
		UNCOMMON = 1;
		RARE = 2;
		EPIC = 3;
		LEGENDARY = 4;
	}
	
	enum EquipType {
		WEAPON = 0;
		ARMOR = 1;
		ACCESSORY = 2;
	}
	
	enum ClassType {
		WARRIOR = 0;
		ARCHER = 1;
		MAGE = 2;	
	}
}

//------------------------------------------------------------------------------------------

message FullUserStructureProto {
	required int32 id = 1;
	required int32 userId = 2;
	required int32 structId = 3;
	optional int64 lastRetrieved = 4;
	required CoordinateProto coordinates = 5;
	required int32 level = 6;
	required int64 purchaseTime = 7;
	required bool isComplete = 8;
}

//------------------------------------------------------------------------------------------

message FullUserEquipProto {
	required int32 userId = 1;
	required int32 equipId = 2;
	required int32 quantity = 3;
	required bool isStolen = 4;
}


//------------------------------------------------------------------------------------------

message FullTaskProto {
	required int32 taskId = 1;
	required string name = 2;
	required int32 cityId = 3;
	required int32 numRequiredForCompletion = 4;
	required int32 energyCost = 5;
	required int32 minCoinsGained = 6;
	required int32 maxCoinsGained = 7;
	required float chanceOfEquipLoot = 8;
	repeated int32 potentialLootEquipIds = 9;
	required int32 expGained = 10;
	required int32 assetNumWithinCity = 11;
}

//------------------------------------------------------------------------------------------

message CoordinateProto {
	required int32 x = 1;
	required int32 y = 2;
}

//------------------------------------------------------------------------------------------

message LocationProto {
	required float latitude = 1;
	required float longitude = 2;
}

//------------------------------------------------------------------------------------------

enum MarketplacePostType {
	EQUIP_POST = 0;
	WOOD_POST = 1;
	DIAMOND_POST = 2;
	COIN_POST = 3;
}

message FullMarketplacePostProto {
	required int32 marketplacePostId = 1;
	required int32 posterId = 2;
	required MarketplacePostType postType = 3;
	required int64 timeOfPost = 4;
	optional FullEquipProto postedEquip = 5;
	optional int32 postedWood = 6;
	optional int32 postedDiamonds = 7;
	optional int32 postedCoins = 8;
	optional int32 diamondCost = 9;
	optional int32 coinCost = 10;
	optional int32 woodCost = 11;
}

//------------------------------------------------------------------------------------------

message FullStructureProto { 

}

//------------------------------------------------------------------------------------------

enum MarketplaceJobRequirementType { 
	BUY = 0;
	SELL = 1;
}

//------------------------------------------------------------------------------------------

enum CritStructType {
	AVIARY = 0;
	LUMBERMILL = 1;
	CARPENTER = 2;
	VAULT = 3;
	ARMORY = 4;
	MARKETPLACE = 5;
}

//------------------------------------------------------------------------------------------

message FullUserQuestDataLarge {
     required int32 userId = 1;
     required int32 questId = 2;
     required bool complete = 3;
     repeated MinimumUserTaskProto requiredTasksProgress = 4;
     repeated MinimumUserDefeatTypeJobProto requiredDefeatTypeJobProgress = 5;
     repeated MinimumUserBuildStructJobProto requiredBuildStructJobProgress = 6;
     repeated MinimumUserMarketplaceJobProto requiredMarketplaceJobProgress = 7;
     repeated MinimumUserUpgradeStructJobProto requiredUpgradeStructJobProgress = 8;
     repeated MinimumUserPossessEquipJobProto requiredPossessEquipJobProgress = 9;
}

//------------------------------------------------------------------------------------------

message MinimumUserTaskProto {
     required int32 userId = 1;
     required int32 taskId = 2;
     required int32 numTimesActed = 3;
     required int32 numTimesRequired = 4;          //we could replace taskId with a FullTaskProto, which also has numReq
}

//------------------------------------------------------------------------------------------

message MinimumUserDefeatTypeJobProto {
     required int32 userId = 1;
     required int32 questId = 2;
     required DefeatTypeJobProto defeatTypeJobId = 3;
     required int32 numDefeated = 4;
}

//------------------------------------------------------------------------------------------

message DefeatTypeJobProto {
     required int32 id = 1;
     required UserType typeOfEnemy = 2;
     required int32 numEnemiesToDefeat = 3;
     required int32 cityId = 4;
}

//------------------------------------------------------------------------------------------

message MinimumUserBuildStructJobProto {
     required int32 userId = 1;
     required int32 questId = 2;
     required BuildStructJobProto buildStructJob = 3;
     required int32 numOfStructUserHas = 4;
}

//------------------------------------------------------------------------------------------

message BuildStructJobProto {
     required int32 id = 1;
     required int32 structId = 2;
     required int32 quantityRequired = 3;
}

//------------------------------------------------------------------------------------------

message MinimumUserMarketplaceJobProto {
     required int32 userId = 1;
     required int32 questId = 2;
     required MarketplaceJobProto marketplaceJob = 3;
     required int32 numTransactionsDone = 4;
}

//------------------------------------------------------------------------------------------

message MarketplaceJobProto {
     required int32 id = 1;
     required MarketplaceJobRequirementType type = 2;
     required int32 quantityRequired = 3;
}

//------------------------------------------------------------------------------------------

message MinimumUserUpgradeStructJobProto {
     required int32 userId = 1;
     required int32 questId = 2;
     required UpgradeStructJobProto upgradeStructJob = 3;
     required bool isComplete = 4;
}

//------------------------------------------------------------------------------------------

message UpgradeStructJobProto {
     required int32 id = 1;
     required int32 structId = 2;
     required int32 levelReq = 3;
}

//------------------------------------------------------------------------------------------

message MinimumUserPossessEquipJobProto {
     required int32 userId = 1;
     required int32 questId = 2;
     required PossessEquipJobProto possessEquipJobProto = 3;
     required bool isComplete = 4;
}

//------------------------------------------------------------------------------------------

message PossessEquipJobProto {
     required int32 id = 1;
     required int32 equipId = 2;
     required int32 quantityReq = 3;
}

//------------------------------------------------------------------------------------------

message FullQuestProto {
     required int32 id = 1;
     required int32 cityId = 2;
     required string name = 3;
     required string description = 4;
     required string doneResponse = 5;
     required string inProgress = 6;
     required int32 assetNumWithinCity = 7;
     optional int32 coinsGained = 8;
     optional int32 diamondsGained = 9;
     optional int32 woodGained = 10;
     optional int32 expGained = 11;
     optional int32 equipIdGained = 12;
     repeated int32 questsRequiredForThis = 13;
     repeated int32 taskReqs = 14;
     repeated int32 upgradeStructJobsReqs = 15;
     repeated int32 buildStructJobsReqs = 16;
     repeated int32 marketplaceJobsReqs = 17;
     repeated int32 defeatTypeReqs = 18;
     repeated int32 possessEquipJobReqs = 19;
}